File: terminator-application/src/test/java/io/quartic/terminator/CollectingEndpoint.java
package io.quartic.terminator;

import com.google.common.collect.ImmutableList;

import javax.websocket.Endpoint;
import javax.websocket.EndpointConfig;
import javax.websocket.MessageHandler;
import javax.websocket.Session;
import java.io.IOException;
import java.util.List;
import java.util.concurrent.TimeUnit;

import static com.google.common.collect.Lists.newArrayList;
import static io.quartic.common.serdes.ObjectMappers.OBJECT_MAPPER;

public class CollectingEndpoint<T> extends Endpoint {
    private final Class<? extends T> clazz;
    private final List<T> messages = newArrayList();

    public CollectingEndpoint(Class<? extends T> clazz) {
        this.clazz = clazz;
    }

    @Override
    public void onOpen(Session session, EndpointConfig config) {
        // Suppress conversion because this somehow borks the websocket internals - "unable to find decoder"
        //noinspection Convert2Lambda
        session.addMessageHandler(new MessageHandler.Whole<String>() {
            @Override
            public void onMessage(String message) {
                try {
                    messages.add(OBJECT_MAPPER.readValue(message, clazz));
                } catch (IOException e) {
                    throw new RuntimeException(e);
                }
            }
        });
    }

    public boolean awaitMessages(int expected, long timeout, TimeUnit unit) {
        while (timeout != 0 && messages.size() < expected) {
            try {
                unit.sleep(1);
            } catch (InterruptedException e) {
                throw new IllegalStateException("Interrupted", e);
            }
            timeout--;
        }
        return messages.size() >= expected;
    }

    public List<T> messages() {
        return ImmutableList.copyOf(messages);
    }
}

File: terminator-application/src/test/kotlin/io/quartic/terminator/CollectingEndpoint.kt
package io.quartic.terminator

import io.quartic.common.serdes.OBJECT_MAPPER
import java.util.concurrent.TimeUnit
import javax.websocket.Endpoint
import javax.websocket.EndpointConfig
import javax.websocket.MessageHandler.Whole
import javax.websocket.Session

class CollectingEndpoint<out T>(private val converter: (String) -> T) : Endpoint() {
    companion object {
        inline fun <reified T : Any> create(): CollectingEndpoint<T> {
            return CollectingEndpoint({ OBJECT_MAPPER.readValue(it, T::class.java) })
        }
    }

    private val _messages = mutableListOf<T>()
    val messages: List<T> get() = _messages.toList()

    override fun onOpen(session: Session, config: EndpointConfig) =
            session.addMessageHandler(Whole<String> { message -> _messages.add(converter(message)) })

    fun awaitMessages(expected: Int, timeout: Long, unit: TimeUnit): Boolean {
        var tick = timeout
        while (tick != 0L && _messages.size < expected) {
            unit.sleep(1)
            tick--
        }
        return _messages.size >= expected
    }
}
